### ISSUE TITLE
First install, storefront network error

### ISSUE THREAD

**Issue by blackyzero** [medusajs/medusa#446]
Hello,

i did a fresh installation, follow the instruction, all good, no error.
First launch storefront using nextjs, and shows following error.

`Unhandled Runtime Error

Error: Network Error
Call Stack
createError
node_modules/axios/lib/core/createError.js (16:0)
handleError
node_modules/axios/lib/adapters/xhr.js (99:0)`

The attached screenshot shows as below.
How could i fix the issue? Thank you.

https://ibb.co/X4DKmvV

**Comment by kasperkristensen** [comment]
Hi @blackyzero,

It is difficult to say without seeing what request is causing the error, you can see this in networks tab in your browsers developer tools. A guess would be that the error is due to a CORS error. Can you verify that the `const STORE_CORS` is set to `localhost:8000` in your [medusa-config.js](https://github.com/medusajs/medusa-starter-default/blob/master/medusa-config.js) in the `medusa-starter-default` project :)

**Comment by srindom** [comment]
Hi @blackyzero thanks for trying out Medusa! 

As @kasperkristensen is pointing out our default store cors is `localhost:8000`, however, it looks like your storefront is running on IP `10.0.0.38`. That case you have to update your `STORE_CORS` in `backend/medusa-config.js` to `http://10.0.0.38:8000`. 

Also note that if your Medusa server is running on a different IP than localhost you should update your storefront to make requests to that IP as well. If necessary you can update the env var `NEXT_PUBLIC_MEDUSA_URL` to the address where your Medusa server is running.

**Comment by blackyzero** [comment]
Thanks Both your assist,
I tried to change the binding ip to 10.0.0.38, and relaunch the command, but still having same issue.

`// CORS when consuming Medusa from admin
const ADMIN_CORS = process.env.ADMIN_CORS || "http://10.0.0.38:7000";

// CORS to avoid issues when consuming Medusa from a client
const STORE_CORS = process.env.STORE_CORS || "http://10.0.0.38:8000";`

Is there anything that i still missed?
Thanks again.

**Comment by blackyzero** [comment]
Just giving update here. After replacing "localhost" by "10.0.0.38" and the issue is gone.
`root@medusa:/opt/services/my-medusa-store/my-medusa-storefront/utils# cat client.js
import Medusa from "@medusajs/medusa-js";

const BACKEND_URL =
  process.env.NEXT_PUBLIC_MEDUSA_URL || "http://10.0.0.38:9000";

export const createClient = () => new Medusa({ baseUrl: BACKEND_URL });
root@medusa:/opt/services/my-medusa-store/my-medusa-storefront/utils#`

Thanks all.

**Comment by javimey** [comment]
Hi,

I cannot make medusa to work with POST requests in my storefront. I can perform a GET request to retrieve products but not the post to generate the cart: 

```
Access to XMLHttpRequest at 'http://localhost:9000/store/carts' from origin 'http://localhost:8000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.
```

I configured medusa-config.js as

```
const STORE_CORS = process.env.STORE_CORS || "http://localhost:8000"
```

And get request go through (like the one for products).

Any ideas?

**Comment by olivermrbl** [comment]
> I can perform a GET request to retrieve products but not the post to generate the cart

Does your server log or browser network tab give you an indication of what might be wrong? It's not a CORS issue since your GET requests are successful, so I am thinking that it might be related to something else.

**Comment by javimey** [comment]
Hi @olivermrbl 

I reinstalled the system and I realized that in NextJS the CORS issue was getting triggered, but when I used the Gatsby implementation, it was complaining about other things (Not having a region configured, prices not belonging to the region, etc).

I configured it using the admin frontend.

Thanks for your help!

**Comment by olivermrbl** [comment]
Glad to hear that you got it working!

**Comment by santoni0** [comment]
Hello, today I tried installing medusa, but any of the installation methods work for me. I always get the same error: 
![image](https://user-images.githubusercontent.com/66315787/163273224-3bed23b2-cb3c-46e0-be2d-94a178807125.png)
Console shows:
`Access to XMLHttpRequest at 'http://localhost:9000/store/carts' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`
and:
`POST http://localhost:9000/store/carts net::ERR_FAILED 400`

Could anyone help me fix these errors?

**Comment by olivermrbl** [comment]
It looks like you are experiencing  CORS issues. Check out [this](https://docs.medusajs.com/troubleshooting/cors-issues) troubleshooting section. You need to configure a different CORS for your storefront, since you are running on port 3000, and we default to 8000.

**Comment by santoni0** [comment]
I'd already tried that. Here's my medusa-config.js.
```
// CORS to avoid issues when consuming Medusa from a client
const STORE_CORS = process.env.STORE_CORS || "http://localhost:3000";

```

**Comment by santoni0** [comment]
Yesterday I updated medusa npm package to the dev version, used example storefront, and I still get the same error. Does anyone know a solution for this problem?

**Comment by pasqualesimonetti** [comment]
Same here on MacOS Monterey 12.3.1, Node 17.9.0

Just followed quickstart instructions for the Medusa engine and Medusa Admin.
But I got no errors in console logs about ports.

<img width="1920" alt="Schermata 2022-04-22 alle 12 14 33" src="https://user-images.githubusercontent.com/2115829/164688140-aa66740f-8186-4f7d-b137-445f57ddc969.png">

**Comment by khushal-1704** [comment]
> Same here on MacOS Monterey 12.3.1, Node 17.9.0, Axios 0.24.0
> 
> Just followed quickstart instructions for the Medusa engine and Medusa Admin. But I get no errors in console logs about ports.
> 
> This is what happens when I go to http://localhost:7001/
> 
> <img alt="Schermata 2022-04-22 alle 12 14 33" width="1920" src="https://user-images.githubusercontent.com/2115829/164688140-aa66740f-8186-4f7d-b137-445f57ddc969.png">

facing the same issue, on both admin and storefront

**Comment by Imran-S-heikh** [comment]
If anyone facing the issue, Maybe this is the cache issue, try to delete the site cache or open storefront in incognito mode. Cleaning the cache manually solved the problem for me. 

TIP: Delete cache after changing CORS settings.

**Comment by olivermrbl** [comment]
> Yesterday I updated medusa npm package to the dev version, used example storefront, and I still get the same error. Does anyone know a solution for this problem?

@santoni0  Did you manage to resolve the issue? 

It looks like you are receiving an error 400, which might be related to the bug fixed in this [PR](https://github.com/medusajs/nextjs-starter-medusa/pull/24). Could I get you to try pull in latest changes and restart the storefront?

**Comment by cboydstun** [comment]
The issue appears to be with the environmental variable `GATSBY_MEDUSA_BACKEND_URL` not taking precedence during loading. I overcame this issue in the Admin project in Chrome and FireFox on a *local* development with deployed Medusa backend server:

1. Open Gatsby Admin project 
2. Go to file `config.js` inside `src` directory -> change line 3 to `let medusaUrl = "YOUR DEPLOYED MEDUSA SERVER"`
![image](https://user-images.githubusercontent.com/52934303/173563326-ee7969cf-8084-4d15-b952-8029bb887e42.png)

3. Go to file `request.js` inside `src` directory -> change line 3 to `let baseURL = "YOUR DEPLOYED MEDUSA SERVER"`
![image](https://user-images.githubusercontent.com/52934303/173563431-e7af26af-3768-4966-8b57-33922f2ef8cd.png)


No more issues after changing those two variables and removing `GATSBY_MEDUSA_BACKEND_URL` as an environmental variable from the Admin project! Also, to apply this solution, delete the `ADMIN_CORS` variable from your deployed Medusa server so that it defaults back to the `localhost:7000` that it is expecting. I'll be working on getting it all deployed today so stay tuned! 😃

**Comment by rsurjano** [comment]
Try again cleaning cookies, it works for me (also try in incognito mode)

**Comment by impactcolor** [comment]
I'm getting CORS errors with everything set as indicated on the starter example:
`// CORS when consuming Medusa from admin
const ADMIN_CORS = process.env.ADMIN_CORS || "http://localhost:7000,http://localhost:7001";

// CORS to avoid issues when consuming Medusa from a client
const STORE_CORS = process.env.STORE_CORS || "http://localhost:8000";`
I can confirm both are running on those ports.  The error is the message on the documentation regarding cors.

I also have cleared cache and cookies.  Any idea what's going on?

**Comment by impactcolor** [comment]
I've even tried streaming port 8000 through ngronk
Access to XMLHttpRequest at 'http://localhost:9000/store/carts' from origin 'https://4321-172-114-11-112.ngrok.io/' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
while changing it in the .env file

**Comment by impactcolor** [comment]
after further inspection this ONLY happens on this request.  localhost:9000/products loads what would be the cause of this?

**Comment by hckia** [comment]
I actually ran into this problem too, and it did mimic a CORs issue, and indeed the cause was no region_id. An alternative method that worked for me was to seed data in the backend after install (npm run seed) if you're doing this just for testing. I assumed that was already done, but it seems some important items were missing.

**Comment by matbrgz** [comment]
I'm running an application in a Docker container based on the repository https://github.com/medusajs/docker-medusa. I've encountered an unhandled runtime error with the following call stack:

![image](https://user-images.githubusercontent.com/23284276/230406734-667acfb8-889f-483b-aca1-8519f778f814.png)
```
1 of 1 unhandled error

Unhandled Runtime Error
Error: Request failed with status code 400

Call Stack
createError
node_modules/@medusajs/medusa-js/node_modules/axios/lib/core/createError.js (16:0)
settle
node_modules/@medusajs/medusa-js/node_modules/axios/lib/core/settle.js (17:0)
XMLHttpRequest.onloadend
node_modules/@medusajs/medusa-js/node_modules/axios/lib/adapters/xhr.js (66:0)
```


Dockerfile
```
FROM node:18.15.0
EXPOSE 8000
WORKDIR /app/storefront
COPY package*.json ./
RUN rm -rf node_modules
RUN apt-get update
RUN npm install -g npm@latest
RUN npm install -g next-cli
COPY . .
RUN npm install
RUN npm run build &> /dev/null
COPY develop.sh .
ENTRYPOINT ["./develop.sh", "develop"]
```

develop.sh
```
#!/bin/bash

npm run $1
```

I wanted to report this issue so it can be addressed and resolved. Thank you for your attention to this matter.

**Comment by armohamm** [comment]
I too was receiving "Network Errror" from strorefront running on localhost at port 8000 and no seeded medusa s3 images were loading while in backend(http://localhost:9000/store/products) and admin(http://localhost:9000/admin) was working fine. 
For me It was a region issue as medusa nextjs storefront looks for region_id and country_code  to select currency while startup, 
After seeding default data, I had to add new region and added few new currencies related to my project, Default region being "EU" followed by "NA", just make sure, you have selected as default the right region i.e "NA" or "EU" on top and relevant currency is set to default, and since the seeded data has data with these two regions only, you may encounter error due to having different currencies and selection of region as default, i.e. the one with the top should be the default region i guess. I need to explore more in code here. To be sure, if that's the real culprit to your network error's, make sure to test by selecting the region = "NA" and currency to "US Dollars" as default, it should even work with "EU" as the seeded products are with these region and currencies, unless until you have customized it. Guess, that's one of the reason for "Network Error" at least it worked for me that way. Hope that helps.
Note: use admin from medusa inbuilt admin at port 9000 as shown above instead of downloading from medusa admin repository and running seperately(i.e. http://localhost:7000,http://localhost:7001), at least for new installations.

**Comment by vasilemidrigan** [comment]
I want to share here my solution to the error below as I see it is mentioned here in some comments.

**Access to XMLHttpRequest at 'http://localhost:9000/store/products' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.**

`STORE_CORS ` variable is initially set to `process.env.STORE_CORS || "http://localhost:3000"` which means it will take the value from env variable if that exists and is true, so instead what I did, I just changed STORE_CORS to "localhost:3000"(which is the client from where I make requests to the server). 

By the way, in medusa-config.js it is pretty clear commented, but I somehow didn't payed attention initially.

**So, in your backend, in medusa-config.js file:** 

```
// CORS to avoid issues when consuming Medusa from a client
const STORE_CORS = process.env.STORE_CORS || "http://localhost:8000"; 
```
**change STORE_CORS to your client localhost. In my case it was 3000.** 
```
// CORS to avoid issues when consuming Medusa from a client
const STORE_CORS = "http://localhost:3000"; 
```

Hope that helps!
